File: src/lib/workspace.ts
import type { Artist, Prisma } from "@prisma/client";
import { CapacityFullError, claimCapacitySeats } from "./capacity";
import { BLOCKING_STATUSES, DEFAULT_DAILY_CAPACITY } from "./constants";
import { db } from "./db";
import { defaultCapacityForRoles, hasManagementRole, memberMatchesServiceKind, parseRoles, rolesFromSpecialty } from "./roles";
import { deriveShiftsFromWindow, typicalWindow } from "./shifts";

type Tx = Prisma.TransactionClient | typeof db;

export class WorkspaceError extends Error {
  constructor(
    message: string,
    public status = 400,
  ) {
    super(message);
    this.name = "WorkspaceError";
  }
}

const businessInclude = {
  owner: true,
  members: { include: { services: true, artist: { select: { id: true, name: true, phone: true, slug: true } } } },
  shifts: { orderBy: { sortOrder: "asc" as const } },
  hours: true,
  blocked: true,
  services: { select: { id: true, kind: true, active: true } },
} satisfies Prisma.BusinessInclude;

export type LoadedBusiness = Prisma.BusinessGetPayload<{ include: typeof businessInclude }>;

export type StudioPermissions = {
  canManageBusiness: boolean;
  canManageTeam: boolean;
  canManageServices: boolean;
  canViewFees: boolean;
  canAssign: boolean;
  canSeeBrideContact: boolean;
};

export type Workspace = {
  artist: Artist;
  business: LoadedBusiness;
  member: LoadedBusiness["members"][number];
  permissions: StudioPermissions;
};

function permissionsFor(roles: string[]): StudioPermissions {
  const manage = hasManagementRole(roles);
  return {
    canManageBusiness: manage,
    canManageTeam: manage,
    canManageServices: manage,
    canViewFees: manage,
    canAssign: manage,
    canSeeBrideContact: manage,
  };
}

export async function loadBusiness(tx: Tx, id: string): Promise<LoadedBusiness> {
  return tx.business.findUniqueOrThrow({
    where: { id },
    include: businessInclude,
  });
}

export async function lockBusiness(tx: Tx, businessId: string) {
  await tx.business.update({ where: { id: businessId }, data: { updatedAt: new Date() } });
}

export async function lockTeamMembers(tx: Tx, ids: string[]) {
  for (const id of [...new Set(ids)]) {
    await tx.teamMember.update({ where: { id }, data: { updatedAt: new Date() } });
  }
}

async function uniqueBusinessSlug(tx: Tx, preferred: string, artistId: string) {
  let slug = preferred;
  for (let i = 0; i < 8; i += 1) {
    const taken = await tx.business.findUnique({ where: { slug } });
    if (!taken) return slug;
    slug = `${preferred}-${artistId.slice(-4).toLowerCase()}${i || ""}`;
  }
  return `${preferred}-${Date.now().toString(36)}`;
}

async function defaultShifts(tx: Tx, artist: Artist, businessId: string) {
  const hours = await tx.weeklyHour.findMany({ where: { OR: [{ artistId: artist.id }, { businessId }] } });
  const window = typicalWindow(hours);
  return deriveShiftsFromWindow(window.startMin, window.endMin);
}

export async function syncShiftsFromHours(tx: Tx, artist: Artist, businessId: string) {
  const existing = await tx.shift.findMany({ where: { businessId } });
  if (existing.length) return existing;
  const drafts = await defaultShifts(tx, artist, businessId);
  await tx.shift.createMany({
    data: drafts.map((shift) => ({
      businessId,
      key: shift.key,
      nameAr: shift.nameAr,
      nameEn: shift.nameEn,
      startMin: shift.startMin,
      endMin: shift.endMin,
      sortOrder: shift.sortOrder,
      capacity: null,
      active: true,
    })),
  });
  return tx.shift.findMany({ where: { businessId }, orderBy: { sortOrder: "asc" } });
}

async function attachExistingRows(tx: Tx, artist: Artist, businessId: string, memberId: string) {
  await tx.weeklyHour.updateMany({ where: { artistId: artist.id, businessId: null }, data: { businessId } });
  await tx.blockedDate.updateMany({ where: { artistId: artist.id, businessId: null }, data: { businessId } });
  await tx.service.updateMany({ where: { artistId: artist.id, businessId: null }, data: { businessId } });
  await tx.booking.updateMany({ where: { artistId: artist.id, businessId: null }, data: { businessId } });
  await tx.platformFee.updateMany({ where: { artistId: artist.id, businessId: null }, data: { businessId } });

  const services = await tx.service.findMany({ where: { businessId } });
  for (const service of services) {
    const assigned = await tx.teamMemberService.count({ where: { serviceId: service.id } });
    if (assigned > 0) continue;
    await tx.teamMemberService.upsert({
      where: { teamMemberId_serviceId: { teamMemberId: memberId, serviceId: service.id } },
      update: {},
      create: { teamMemberId: memberId, serviceId: service.id },
    });
  }

  const shifts = await tx.shift.findMany({ where: { businessId, active: true } });
  const bookings = await tx.booking.findMany({
    where: { businessId },
    include: { items: true, assignments: true, capacityHolds: true },
  });
  const member = await tx.teamMember.findUnique({ where: { id: memberId } });
  for (const booking of bookings) {
    if (!booking.assignments.length) {
      const serviceIds = booking.items.length ? booking.items.map((item) => item.serviceId) : [booking.serviceId];
      for (const serviceId of serviceIds) {
        await tx.bookingAssignment.upsert({
          where: { bookingId_serviceId: { bookingId: booking.id, serviceId } },
          update: {},
          create: { bookingId: booking.id, teamMemberId: memberId, serviceId },
        });
      }
    }
    const covering = shifts.filter((shift) => booking.startMin >= shift.startMin && booking.startMin < shift.endMin);
    const shiftId = covering[0]?.id || shifts[0]?.id || null;
    if (!booking.shiftId && shiftId && booking.scheduleMode !== "HOURLY") {
      await tx.booking.update({ where: { id: booking.id }, data: { shiftId } });
    }
    if (
      member &&
      (BLOCKING_STATUSES as readonly string[]).includes(booking.status) &&
      booking.capacityHolds.length === 0
    ) {
      try {
        await claimCapacitySeats(tx, {
          date: booking.date,
          shiftId,
          bookingId: booking.id,
          memberIds: [memberId],
          members: [{ id: member.id, dailyCapacity: member.dailyCapacity }],
          shiftCapacity: covering[0]?.capacity ?? shifts[0]?.capacity ?? null,
        });
      } catch (error) {
        if (!(error instanceof CapacityFullError)) throw error;
      }
    }
  }
}

export async function createOwnedBusiness(tx: Tx, artist: Artist, input?: { name?: string; businessType?: string; slug?: string }) {
  const slug = await uniqueBusinessSlug(tx, input?.slug || artist.slug, artist.id);
  const business = await tx.business.create({
    data: {
      ownerId: artist.id,
      name: input?.name?.trim() || artist.name,
      slug,
      businessType: input?.businessType === "salon" ? "salon" : "independent",
      scheduleMode: "SHIFT",
      assignmentMode: "AUTO",
      neighborhood: artist.neighborhood,
      city: artist.city,
      phone: artist.phone,
      bio: artist.bio,
      bookingHorizonDays: artist.bookingHorizonDays,
      minNoticeHours: artist.minNoticeHours,
    },
  });
  const member = await tx.teamMember.create({
    data: {
      businessId: business.id,
      artistId: artist.id,
      name: artist.name,
      phone: artist.phone,
      roles: rolesFromSpecialty(artist.specialty),
      dailyCapacity: DEFAULT_DAILY_CAPACITY,
      status: "ACTIVE",
    },
  });
  await syncShiftsFromHours(tx, artist, business.id);
  await attachExistingRows(tx, artist, business.id, member.id);
  return business.id;
}

async function loadWorkspaceByMember(tx: Tx, artist: Artist, memberId: string): Promise<Workspace> {
  const member = await tx.teamMember.findUnique({ where: { id: memberId } });
  if (!member || member.status !== "ACTIVE") throw new WorkspaceError("UNAUTHORIZED", 401);
  const business = await tx.business.findUniqueOrThrow({
    where: { id: member.businessId },
    include: businessInclude,
  });
  const hydrated = business.members.find((row) => row.id === member.id);
  if (!hydrated) throw new WorkspaceError("UNAUTHORIZED", 401);
  return {
    artist,
    business,
    member: hydrated,
    permissions: permissionsFor(parseRoles(hydrated.roles)),
  };
}

export async function ensureWorkspace(artist: Artist, tx: Tx = db): Promise<Workspace> {
  const owned = await tx.business.findFirst({ where: { ownerId: artist.id }, orderBy: { createdAt: "asc" } });
  if (owned) {
    const member = await tx.teamMember.findFirst({
      where: { businessId: owned.id, artistId: artist.id, status: "ACTIVE" },
    });
    if (!member) {
      const created = await tx.teamMember.create({
        data: {
          businessId: owned.id,
          artistId: artist.id,
          name: artist.name,
          phone: artist.phone,
          roles: rolesFromSpecialty(artist.specialty),
          dailyCapacity: defaultCapacityForRoles(rolesFromSpecialty(artist.specialty)),
          status: "ACTIVE",
        },
      });
      await attachExistingRows(tx, artist, owned.id, created.id);
      await syncShiftsFromHours(tx, artist, owned.id);
      return loadWorkspaceByMember(tx, artist, created.id);
    }
    await syncShiftsFromHours(tx, artist, owned.id);
    await attachExistingRows(tx, artist, owned.id, member.id);
    return loadWorkspaceByMember(tx, artist, member.id);
  }

  const invited = await tx.teamMember.findFirst({
    where: {
      status: "ACTIVE",
      OR: [{ artistId: artist.id }, { phone: artist.phone, artistId: null }],
    },
    orderBy: { createdAt: "asc" },
  });
  if (invited) {
    if (!invited.artistId) {
      await tx.teamMember.update({
        where: { id: invited.id },
        data: { artistId: artist.id, name: invited.name || artist.name, phone: artist.phone },
      });
    }
    return loadWorkspaceByMember(tx, artist, invited.id);
  }

  const businessId = await createOwnedBusiness(tx, artist);
  const member = await tx.teamMember.findFirstOrThrow({
    where: { businessId, artistId: artist.id },
  });
  return loadWorkspaceByMember(tx, artist, member.id);
}

export async function requireWorkspace(artist: Artist) {
  return ensureWorkspace(artist);
}

export function requirePermission(workspace: Workspace, key: keyof StudioPermissions) {
  if (!workspace.permissions[key]) {
    throw new WorkspaceError("FORBIDDEN", 403);
  }
}

export async function migrateAllArtists(tx: Tx = db) {
  const artists = await tx.artist.findMany();
  for (const artist of artists) {
    await ensureWorkspace(artist, tx);
  }
}

export async function findBusinessBySlug(slug: string, tx: Tx = db) {
  const direct = await tx.business.findUnique({
    where: { slug },
    include: businessInclude,
  });
  if (direct) return direct;
  const artist = await tx.artist.findUnique({ where: { slug } });
  if (!artist) return null;
  const owned = await tx.business.findFirst({ where: { ownerId: artist.id } });
  if (!owned && !artist.onboardingComplete) return null;
  const workspace = await ensureWorkspace(artist, tx);
  if (workspace.business.ownerId !== artist.id) return null;
  return workspace.business;
}

export function activeMembers(business: LoadedBusiness) {
  return business.members.filter((member) => member.status === "ACTIVE");
}

export function memberCanPerform(member: LoadedBusiness["members"][number], serviceId: string, kind?: string) {
  if (kind) return memberMatchesServiceKind(parseRoles(member.roles), kind);
  return member.services.some((row) => row.serviceId === serviceId);
}

export async function syncServiceStaffByRole(tx: Tx, businessId: string) {
  const [members, services] = await Promise.all([
    tx.teamMember.findMany({ where: { businessId, status: "ACTIVE" } }),
    tx.service.findMany({ where: { businessId }, include: { staff: true } }),
  ]);
  for (const service of services) {
    const eligibleIds = new Set(
      members.filter((member) => memberMatchesServiceKind(parseRoles(member.roles), service.kind)).map((member) => member.id),
    );
    const currentIds = new Set(service.staff.map((row) => row.teamMemberId));
    const toAdd = [...eligibleIds].filter((id) => !currentIds.has(id));
    const toRemove = [...currentIds].filter((id) => !eligibleIds.has(id));
    if (toRemove.length) {
      await tx.teamMemberService.deleteMany({
        where: { serviceId: service.id, teamMemberId: { in: toRemove } },
      });
    }
    if (toAdd.length) {
      await tx.teamMemberService.createMany({
        data: toAdd.map((teamMemberId) => ({ teamMemberId, serviceId: service.id })),
      });
    }
  }
}

export { isTeamBusiness } from "./roles";

export async function promoteToSalonIfTeam(tx: Tx, businessId: string) {
  const business = await tx.business.findUniqueOrThrow({
    where: { id: businessId },
    select: { businessType: true },
  });
  if (business.businessType === "salon") return;
  const activeCount = await tx.teamMember.count({
    where: { businessId, status: "ACTIVE" },
  });
  if (activeCount > 1) {
    await tx.business.update({
      where: { id: businessId },
      data: { businessType: "salon" },
    });
  }
}

export function bookingScopeWhere(workspace: Workspace): Prisma.BookingWhereInput {
  if (workspace.permissions.canManageBusiness) {
    return { businessId: workspace.business.id };
  }
  return {
    businessId: workspace.business.id,
    assignments: { some: { teamMemberId: workspace.member.id } },
  };
}

File: backend/bridey_api/workspace.py
from __future__ import annotations

from datetime import datetime, timezone

from sqlalchemy import or_, select, update
from sqlalchemy.orm import Session, selectinload

from bridey_api.capacity import claim_capacity_seats
from bridey_api.constants import BLOCKING_STATUSES, DEFAULT_DAILY_CAPACITY
from bridey_api.errors import CapacityFullError, WorkspaceError
from bridey_api.ids import new_id
from bridey_api.models import (
    Artist,
    BlockedDate,
    Booking,
    BookingAssignment,
    Business,
    PlatformFee,
    Service,
    Shift,
    TeamMember,
    TeamMemberService,
    WeeklyHour,
)
from bridey_api.roles import default_capacity_for_roles, has_management_role, member_matches_service_kind, parse_roles, roles_from_specialty
from bridey_api.shifts import derive_shifts_from_window, typical_window

BUSINESS_LOAD = (
    selectinload(Business.owner),
    selectinload(Business.members).selectinload(TeamMember.services),
    selectinload(Business.members).selectinload(TeamMember.artist),
    selectinload(Business.shifts),
    selectinload(Business.hours),
    selectinload(Business.blocked),
    selectinload(Business.services),
)


def permissions_for(roles: list[str]) -> dict:
    manage = has_management_role(roles)
    return {
        "canManageBusiness": manage,
        "canManageTeam": manage,
        "canManageServices": manage,
        "canViewFees": manage,
        "canAssign": manage,
        "canSeeBrideContact": manage,
    }


def load_business(session: Session, business_id: str) -> Business:
    business = session.scalars(select(Business).where(Business.id == business_id).options(*BUSINESS_LOAD)).one()
    business.shifts.sort(key=lambda row: row.sort_order)
    return business


def unique_business_slug(session: Session, preferred: str, artist_id: str) -> str:
    slug = preferred
    for i in range(8):
        taken = session.scalar(select(Business).where(Business.slug == slug))
        if not taken:
            return slug
        slug = f"{preferred}-{artist_id[-4:].lower()}{'' if i == 0 else i}"
    return f"{preferred}-{int(datetime.now(timezone.utc).timestamp() * 1000):x}"


def sync_shifts_from_hours(session: Session, artist: Artist, business_id: str) -> list[Shift]:
    existing = session.scalars(select(Shift).where(Shift.business_id == business_id)).all()
    if existing:
        return sorted(existing, key=lambda row: row.sort_order)
    hours = session.scalars(
        select(WeeklyHour).where(or_(WeeklyHour.artist_id == artist.id, WeeklyHour.business_id == business_id))
    ).all()
    window = typical_window(hours)
    drafts = derive_shifts_from_window(window["startMin"], window["endMin"])
    for draft in drafts:
        session.add(
            Shift(
                id=new_id(),
                business_id=business_id,
                key=draft["key"],
                name_ar=draft["nameAr"],
                name_en=draft["nameEn"],
                start_min=draft["startMin"],
                end_min=draft["endMin"],
                sort_order=draft["sortOrder"],
                capacity=None,
                active=True,
            )
        )
    session.flush()
    return session.scalars(select(Shift).where(Shift.business_id == business_id).order_by(Shift.sort_order.asc())).all()


def attach_existing_rows(session: Session, artist: Artist, business_id: str, member_id: str) -> None:
    session.execute(update(WeeklyHour).where(WeeklyHour.artist_id == artist.id, WeeklyHour.business_id.is_(None)).values(business_id=business_id))
    session.execute(update(BlockedDate).where(BlockedDate.artist_id == artist.id, BlockedDate.business_id.is_(None)).values(business_id=business_id))
    session.execute(update(Service).where(Service.artist_id == artist.id, Service.business_id.is_(None)).values(business_id=business_id))
    session.execute(update(Booking).where(Booking.artist_id == artist.id, Booking.business_id.is_(None)).values(business_id=business_id))
    session.execute(update(PlatformFee).where(PlatformFee.artist_id == artist.id, PlatformFee.business_id.is_(None)).values(business_id=business_id))
    session.flush()

    services = session.scalars(select(Service).where(Service.business_id == business_id)).all()
    for service in services:
        assigned = session.scalar(
            select(TeamMemberService).where(TeamMemberService.service_id == service.id)
        )
        if assigned:
            continue
        session.merge(TeamMemberService(team_member_id=member_id, service_id=service.id))

    shifts = session.scalars(select(Shift).where(Shift.business_id == business_id, Shift.active.is_(True))).all()
    bookings = session.scalars(
        select(Booking)
        .where(Booking.business_id == business_id)
        .options(selectinload(Booking.items), selectinload(Booking.assignments), selectinload(Booking.capacity_holds))
    ).all()
    member = session.get(TeamMember, member_id)
    for booking in bookings:
        if not booking.assignments:
            service_ids = [item.service_id for item in booking.items] if booking.items else [booking.service_id]
            for service_id in service_ids:
                exists = session.scalar(
                    select(BookingAssignment).where(
                        BookingAssignment.booking_id == booking.id,
                        BookingAssignment.service_id == service_id,
                    )
                )
                if not exists:
                    session.add(
                        BookingAssignment(
                            id=new_id(),
                            booking_id=booking.id,
                            team_member_id=member_id,
                            service_id=service_id,
                        )
                    )
        covering = [shift for shift in shifts if booking.start_min >= shift.start_min and booking.start_min < shift.end_min]
        shift_id = (covering[0].id if covering else None) or (shifts[0].id if shifts else None)
        if not booking.shift_id and shift_id and booking.schedule_mode != "HOURLY":
            booking.shift_id = shift_id
        if member and booking.status in BLOCKING_STATUSES and not booking.capacity_holds:
            try:
                claim_capacity_seats(
                    session,
                    date=booking.date,
                    shift_id=shift_id,
                    booking_id=booking.id,
                    member_ids=[member_id],
                    members=[{"id": member.id, "dailyCapacity": member.daily_capacity}],
                    shift_capacity=(covering[0].capacity if covering else None) or (shifts[0].capacity if shifts else None),
                )
            except CapacityFullError:
                pass
    session.flush()


def create_owned_business(session: Session, artist: Artist, input: dict | None = None) -> str:
    payload = input or {}
    slug = unique_business_slug(session, payload.get("slug") or artist.slug, artist.id)
    now = datetime.now(timezone.utc)
    business = Business(
        id=new_id(),
        owner_id=artist.id,
        name=(payload.get("name") or "").strip() or artist.name,
        slug=slug,
        business_type="salon" if payload.get("businessType") == "salon" else "independent",
        schedule_mode="SHIFT",
        assignment_mode="AUTO",
        neighborhood=artist.neighborhood,
        city=artist.city,
        phone=artist.phone,
        bio=artist.bio,
        booking_horizon_days=artist.booking_horizon_days,
        min_notice_hours=artist.min_notice_hours,
        created_at=now,
        updated_at=now,
    )
    session.add(business)
    session.flush()
    member = TeamMember(
        id=new_id(),
        business_id=business.id,
        artist_id=artist.id,
        name=artist.name,
        phone=artist.phone,
        roles=roles_from_specialty(artist.specialty),
        daily_capacity=DEFAULT_DAILY_CAPACITY,
        status="ACTIVE",
        created_at=now,
        updated_at=now,
    )
    session.add(member)
    session.flush()
    sync_shifts_from_hours(session, artist, business.id)
    attach_existing_rows(session, artist, business.id, member.id)
    return business.id


def load_workspace_by_member(session: Session, artist: Artist, member_id: str) -> dict:
    member_row = session.get(TeamMember, member_id)
    if not member_row or member_row.status != "ACTIVE":
        raise WorkspaceError("UNAUTHORIZED", 401)
    business = load_business(session, member_row.business_id)
    hydrated = next((row for row in business.members if row.id == member_row.id), None)
    if not hydrated:
        raise WorkspaceError("UNAUTHORIZED", 401)
    return {
        "artist": artist,
        "business": business,
        "member": hydrated,
        "permissions": permissions_for(parse_roles(hydrated.roles)),
    }


def ensure_workspace(session: Session, artist: Artist) -> dict:
    owned = session.scalar(select(Business).where(Business.owner_id == artist.id).order_by(Business.created_at.asc()))
    if owned:
        member = session.scalar(
            select(TeamMember).where(
                TeamMember.business_id == owned.id,
                TeamMember.artist_id == artist.id,
                TeamMember.status == "ACTIVE",
            )
        )
        if not member:
            now = datetime.now(timezone.utc)
            created = TeamMember(
                id=new_id(),
                business_id=owned.id,
                artist_id=artist.id,
                name=artist.name,
                phone=artist.phone,
                roles=roles_from_specialty(artist.specialty),
                daily_capacity=default_capacity_for_roles(roles_from_specialty(artist.specialty)),
                status="ACTIVE",
                created_at=now,
                updated_at=now,
            )
            session.add(created)
            session.flush()
            attach_existing_rows(session, artist, owned.id, created.id)
            sync_shifts_from_hours(session, artist, owned.id)
            return load_workspace_by_member(session, artist, created.id)
        sync_shifts_from_hours(session, artist, owned.id)
        attach_existing_rows(session, artist, owned.id, member.id)
        return load_workspace_by_member(session, artist, member.id)

    invited = session.scalar(
        select(TeamMember)
        .where(
            TeamMember.status == "ACTIVE",
            or_(TeamMember.artist_id == artist.id, (TeamMember.phone == artist.phone) & TeamMember.artist_id.is_(None)),
        )
        .order_by(TeamMember.created_at.asc())
    )
    if invited:
        if not invited.artist_id:
            invited.artist_id = artist.id
            invited.name = invited.name or artist.name
            invited.phone = artist.phone
            session.flush()
        return load_workspace_by_member(session, artist, invited.id)

    business_id = create_owned_business(session, artist)
    member = session.scalars(
        select(TeamMember).where(TeamMember.business_id == business_id, TeamMember.artist_id == artist.id)
    ).one()
    return load_workspace_by_member(session, artist, member.id)


def require_workspace(session: Session, artist: Artist) -> dict:
    return ensure_workspace(session, artist)


def sync_service_staff_by_role(session: Session, business_id: str) -> None:
    members = session.scalars(select(TeamMember).where(TeamMember.business_id == business_id, TeamMember.status == "ACTIVE")).all()
    services = session.scalars(
        select(Service).where(Service.business_id == business_id).options(selectinload(Service.staff))
    ).all()
    for service in services:
        eligible = {member.id for member in members if member_matches_service_kind(parse_roles(member.roles), service.kind)}
        current = {row.team_member_id for row in service.staff}
        to_add = eligible - current
        to_remove = current - eligible
        if to_remove:
            for row in list(service.staff):
                if row.team_member_id in to_remove:
                    session.delete(row)
        for team_member_id in to_add:
            session.add(TeamMemberService(team_member_id=team_member_id, service_id=service.id))
    session.flush()


def lock_business(session: Session, business_id: str) -> None:
    business = session.get(Business, business_id)
    if business:
        business.updated_at = datetime.now(timezone.utc)
        session.flush()


def lock_team_members(session: Session, ids: list[str]) -> None:
    now = datetime.now(timezone.utc)
    for member_id in dict.fromkeys(ids):
        member = session.get(TeamMember, member_id)
        if member:
            member.updated_at = now
    session.flush()


def find_business_by_slug(session: Session, slug: str):
    direct = session.scalars(select(Business).where(Business.slug == slug).options(*BUSINESS_LOAD)).first()
    if direct:
        direct.shifts.sort(key=lambda row: row.sort_order)
        return direct
    artist = session.scalar(select(Artist).where(Artist.slug == slug))
    if not artist:
        return None
    owned = session.scalar(select(Business).where(Business.owner_id == artist.id))
    if not owned and not artist.onboarding_complete:
        return None
    ws = ensure_workspace(session, artist)
    if ws["business"].owner_id != artist.id:
        return None
    return load_business(session, ws["business"].id)
